test/test_functions.hpp

```diff
@@ -0,0 +1,159 @@
+#pragma once
+#include <cmath>
+#include <random>
+#include "quad_matcher.hpp"
+
+
+namespace detail {
+  constexpr size_t factorial(size_t n) {
+    return (n == 1 || n == 0) ? 1 : factorial(n - 1) * n;
+  }
+}
+struct AssociatedLegendre {
+  static inline double evaluate(int l, int m, double x) {
+    return std::assoc_legendre(l, m, x);
+  }
+};
+
+struct SphericalHarmonic {
+  static auto evaluate(int l, int m, double theta, double phi) {
+    double fac_ratio  = detail::factorial(l-m);
+           fac_ratio /= detail::factorial(l+m);
+    double prefactor  = (2.0 * l + 1.0) / (4.0 * M_PI);
+           prefactor *= fac_ratio;
+    prefactor = std::sqrt(prefactor);
+    if( m < 0 ) {
+      prefactor *= std::pow(1, std::abs(m)) * fac_ratio;
+    }
+
+    return prefactor * 
+           std::assoc_legendre(l, std::abs(m), std::cos(theta)) *
+           std::complex( std::cos(m*phi), std::sin(m*phi) );
+  }
+
+  static auto evaluate(int l, int m, double x, double y, double z) {
+    const auto r     = std::sqrt( x*x + y*y + z*z );
+    const auto theta = std::acos( z / r );
+    const auto phi   = std::atan2( y, x );
+
+    return evaluate(l, m, theta, phi);
+  }
+
+  static auto evaluate(int l, int m, std::array<double,3> x) {
+    return evaluate(l, m, x[0], x[1], x[2]);
+  }
+};
+
+struct MagnitudeSquaredSphericalHarmonic {
+  template <typename... Args>
+  static double evaluate( Args&&... args ) {
+    auto eval = std::abs(SphericalHarmonic::evaluate(args...));
+    return eval * eval;
+  }
+};
+
+struct RadialGaussian {
+  static double evaluate(double r) {
+    return r*r * std::exp(-r*r);
+  }
+  static double evaluate(double x, double y, double z) {
+    const auto r = std::hypot(x,y,z);
```

Comment:
Wow there's a function for this? You learn something new every day..

Reply:
Somewhere buried deep in the ISO C++ standard, there is a STL function to wash your car (templated over make and model, of course).